### PR TITLE
fix:  insert critical overflow checks preventing heap corruption

### LINKED ISSUE #3068: Bug: Candid deserialization of large arrays fails with buffer overflow on read
This was hard to reproduce locally using drun (perhaps because of my VM's resources) and annoying to repro with dfx because it refused to upgrade to code with the same hash.

But the playground was just right:

https://m7sm4-2iaaa-aaaab-qabra-cai.raw.ic0.app/?tag=336007141

Consider the code:

```
import Array "mo:base/Array";
import Debug "mo:base/Debug";

actor {
  stable var a : [var Nat] =
    Array.init(268435456 / 2, 0x04); // ca. 0.5GB array
  system func preupgrade() { Debug.print(debug_show({pre=a.size()})); };
  system func postupgrade() { Debug.print(debug_show({post=a.size()})); }
}
```

Deploying and then doing a self-upgrade produces the following crash (see the playground log)

```
[4:42:56 PM] moc version 0.6.17
[4:42:56 PM] base library version dfx-0.8.5-beta.0
[4:43:09 PM] Compiling code...
[4:43:10 PM] Compiled Wasm size: 168KB
[4:43:10 PM] Deploying code...
[4:43:10 PM] Requesting a new canister id...
[4:43:19 PM] Got canister id 4r7kv-yiaaa-aaaab-qac5a-cai
[4:43:29 PM] Code installed at canister id 4r7kv-yiaaa-aaaab-qac5a-cai
[4:43:37 PM] Compiling code...
[4:43:38 PM] Compiled Wasm size: 168KB
[4:43:38 PM] Deploying code...
[4:44:01 PM] Call was rejected: Request ID: 02b62063686f0338440239d22696a69e5be3f1833d36ca63a51b1423deaf0134 Reject code:  Reject text: Canister 4r7kv-yiaaa-aaaab-qac5a-cai trapped explicitly: IDL error: byte read out of buffer
```

I first noticed the bug locally on master (0.6.20), so the outdated moc version of the playground is not the culprit.

This seems to be the line that is trapping:

https://github.com/dfinity/motoko/blob/91df5fd36487633cfb077ae75b807653d8aa4299/rts/motoko-rts/src/buf.rs#L23

Perhaps an error decoding the larger array size and payload range?

UPDATE: Got a drun repro https://github.com/dfinity/motoko/pull/3067  (by decreasing array size)

### PR DESCRIPTION
**Pull request by crusso**
Fixes two classes of `u32`-undetected overflow bugs that could corrupt heap, stable variables, or both:

* ic.rs: `alloc_words` wasn't checking validity of addresses, leading to possible (and observed) heap corruption.
* compile.ml: `buffer_size` wasn't checking overflow of 32-bit size pre-computation.

The first is fixed using an overflow detecting add function of Rust.

The second is checked by using a local `i64` in each type-directed size function. These functions are MV and return two `i32`s (for the data buffer and (vestigial) reference buffer sizes.  I did not modify the functions to return 2 `i64`s because our MultiValue emulation only support `i32` at the moment.

Fixes #3068.

### CODE REVIEW

File: src/codegen/compile.ml [hunk 1]
```diff
@@ -4508,7 +4510,14 @@ module Serialization = struct
         size_alias (fun () -> get_x ^^ Heap.load_field MutBox.field ^^ size env t)
       | _ -> todo "buffer_size" (Arrange_ir.typ t) G.nop
       end ^^
+      (* Check 32-bit overflow of buffer_size *)
       get_data_size ^^
+      compile_const_64 0x1_0000_0000L ^^
```
[the remark this answers is not in the document]
Comment (ggreif) [reply src/codegen/compile.ml:4515]: if the upper bits are `!= 0` then we have an overflow.

**Review by ggreif — APPROVED** [review]
Looks fine! We can do the shift trick in another PR.

File: src/codegen/compile.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (crusso) [reply src/codegen/compile.ml:4515]: Sure, I've done that elsewhere, but is it actually cheaper? I still need to convert the i64 to a i32 for the conditional I think.

**Comment by ggreif** [comment]
No idea. Not urgent.

File: src/codegen/compile.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (crusso) [reply src/codegen/compile.ml]: Done

**Review by ggreif — APPROVED** [review]

**Pull request merged**